examples/sharded-fungible-token/wallet/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+        let refund_to = refund_to.unwrap_or(caller);
+
+        // call receiver's wallet-contract
+        let mut p = Promise::new(state_init.derive_account_id())
+            // refund state_init amount and attached deposit in case of failure
+            // to `refund_to` instead of sender's wallet-contract
+            .refund_to(refund_to.clone());
+
+        if let Some(amount) = state_init_amount {
+            // subtract the required amount for state_init from attached deposit
+            deposit_left = deposit_left
+                .checked_sub(amount)
+                .unwrap_or_else(|| env::panic_str(Self::ERR_INSUFFICIENT_DEPOSIT));
+
+            // deploy & init receiver's wallet-contract
+            p = p.state_init(state_init, amount);
+        }
+
+        // we still need at least 1yN to attach to `.sft_receive()`
+        require!(deposit_left >= NearToken::from_yoctonear(1), Self::ERR_INSUFFICIENT_DEPOSIT);
+
+        Self::ext_on(p)
+            // forward remaining attached deposit
+            .with_attached_deposit(deposit_left)
+            // require minimum gas
+            .with_static_gas(SFTWalletData::SFT_RECEIVE_MIN_GAS)
+            // forward all remaining gas here
+            .with_unused_gas_weight(1)
+            .sft_receive(self.owner_id.clone().into_owned(), amount, memo, notify, Some(refund_to))
+            .then(
+                Self::ext(env::current_account_id())
+                    .with_static_gas(SFTWalletData::SFT_RESOLVE_GAS)
+                    // do not distribute remaining gas here
+                    .with_unused_gas_weight(0)
+                    .sft_resolve(amount, true),
+            )
+            .into()
+    }
+
+    /// Receives tokens from [minter-contract](super::minter::ShardedFungibleTokenMinter)
+    /// or wallet-contracts initialized for the same minter-contract.
+    ///
+    /// If `notify` is set, then `receiver_id::sft_on_transfer()` will be
+    /// called. If `notify.state_init` is set, then `receiver_id` will be
+    /// initialized if doesn't exist.
+    ///
+    /// Remaining attached deposit is refunded to `refund_to` or `sender_id`
+    /// if not set.
+    ///
+    /// Returns `used_amount`.
+    #[payable]
+    fn sft_receive(
+        &mut self,
+        sender_id: AccountId,
+        amount: U128,
+        // memo will be stored in receipt's FunctionCall args anyway
+        #[allow(unused_variables)] memo: Option<String>,
+        notify: Option<TransferNotification>,
+        refund_to: Option<AccountId>,
+    ) -> PromiseOrValue<U128> {
+        let mut deposit_left = env::attached_deposit();
+        require!(deposit_left >= NearToken::from_yoctonear(1), Self::ERR_INSUFFICIENT_DEPOSIT);
+
+        let caller = env::predecessor_account_id();
+        // verify that the caller is a valid wallet-contract or the minter
+        require!(
+            caller == *self.minter_id || caller == self.sft_wallet_account_id_for(&caller),
```

Comment:
Hm, shouldn't this be `self.sft_wallet_account_id_for(&sender_id)`?

```suggestion
            caller == *self.minter_id || caller == self.sft_wallet_account_id_for(&sender_id),
```